hpx/runtime/serialization/variant.hpp

```diff
@@ -0,0 +1,78 @@
+//  Copyright (c) 2015 Anton Bikineev
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef HPX_SERIALIZATION_VARIANT_HPP
+#define HPX_SERIALIZATION_VARIANT_HPP
+
+#include <hpx/runtime/serialization/serialize.hpp>
+
+#include <boost/variant.hpp>
+
+namespace hpx { namespace serialization
+{
+    namespace detail
+    {
+        struct variant_save_visitor: boost::static_visitor<>
+        {
+            variant_save_visitor(output_archive& ar)
+              : ar(ar)
+            {}
+
+            template <class T>
+            void operator()(const T& value)
+            {
+                ar << value;
+            }
+
+        private:
+            output_archive& ar;
+        };
+
+        template <class... Args>
+        struct load_helper_t{};
+
+        template <class Variant, class... Tail>
+        BOOST_FORCEINLINE void load_helper(input_archive& ar, Variant& var,
+                int which, load_helper_t<boost::detail::variant::void_, Tail...>)
```

Comment:
> 1) existing with variadic templates and explicit use of detail::variant::void_;
> I don't see a big problem with use of such implementation detail.

The reason why we use a `detail` namespace is to flag implementation details so users won't touch them. We reserve the right to change them as we see fit, without that being considered a breaking change. Writing our code this way would make us fragile. In fact,  I'm surprised that this works today as there should be no mention of a `void_` anymore since they implemented variadic support. 

> 2) one used in boost.serialization with these macros expanded to `class T1`, ... `class T50` and use of `mpl::list` for recursion.

This again wouldn't work with the variadic version of `variant`, which doesn't have a fixed number of arguments anymore. In that case, this would only match a `variant` of exactly 50 types.

Please add a minimal testcase for this functionality, and I'm sure we'll be able to see it fail with the right combination of compiler and boost version.